### PR TITLE
GitHub actions: store results (#57)

### LINKED ISSUE #38: Use DVC to store results.
As of now, each time we run benchmarks, all benchmarks are run, for all systems. 
After #27 we can identify similar machines.
Use DVC to store the results and `dvc get` them before we run benchmarks.

### LINKED ISSUE #40: Do not build when merging to master
Currently, wokflow is as follows:
1. Prepare PR -> GHA builds it
2. Build successful, review done -> merge to master
3. Another build is triggered, on master this time. In this case only difference between PR build is that we deploy html to gh-pages branch. 

What should happen: After merge, get html from PR build and publish it to gh-pages.

### PR DESCRIPTION
**Pull request by pared**
Fixes #40 (in a way that build will be only on master)
Fixes #38 

Featuring:
- storing and reusing benchmarks results with `dvc`
- "quick build" for pull_requests - PR's from outside of `dvc-bench` repo, are not allowed to have the repo secrets -> no AWS credentials, so actual build should happen after push to `dvc-bench`, we might want to consider "pre-master" branch that would take care of building, so that we are sure master has only "good" changes

### CODE REVIEW

File: run_benchmarks.dvc [hunk 1]
```diff
@@ -1,15 +1,9 @@
-md5: 14267ca8e0770c6290e878383ab37b5a
-cmd: asv run HASHFILE:hashes.txt --profile --show-stderr
+md5: 3cbd61a7c310c11e9d102c5064533fed
+cmd: asv run HASHFILE:hashes.txt -k --profile --show-stderr
```
Comment (pared) [inline run_benchmarks.dvc:2]:
```  
--skip-existing, -k   Skip running benchmarks that have previous successful
or failed results
```

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -1,9 +1,14 @@
...
+  push:
+    branches:
+      - master
 
 jobs:
-  verify:
-    name: tests and styling check
+  build:
+    # Prevent results update PR from triggering the workflow
+    if: github.event.head_commit.author.name != 'github-actions[bot]'
```
Comment (pared) [inline .github/workflows/main.yml:10]: verified [here](https://github.com/iterative/dvc-bench/runs/761168233?check_suite_focus=true)

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -55,32 +56,44 @@ jobs:
...
+        run: dvc pull
       - name: join results
         run: |
-          mkdir results
-          cp -r results-macos-latest/* results/
-          cp -r results-windows-latest/* results/
-          cp -r results-ubuntu-latest/* results/
+          cp -r /tmp/results-windows/* results/
+          cp -r /tmp/results-macos/* results/
+          cp -r /tmp/results-ubuntu/* results/
```
Comment (efiop) [inline .github/workflows/main.yml:85]: Are they on the same fs? If so, could you mv it so it is faster? Though I'm not even sure it takes much time to begin with.

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -55,32 +56,44 @@ jobs:
...
+      - name: commit results to dvc
+        run: dvc commit -f run_benchmarks.dvc && dvc push
```
Comment (efiop) [inline .github/workflows/main.yml:87]: Whoa, this is very neat! :fire: 

Have you looked into run-cache, btw? Seems like we could've made our dvc cache an artifact and then just `dvc pull --run-cache && dvc repro` here instead. Just a question, I really like your current approach too. :ok_hand:

Comment (efiop) [reply .github/workflows/main.yml:87]: Though, with run-cache you won't even have to create a commit/PR. Unless I'm missing something here.

File: .github/workflows/pull_request_build.yml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        run: pip install -r requirements.txt
+      - name: check project styling
+        if: matrix.os == 'ubuntu-latest'
+        run: pre-commit run --all-files
+      - name: run tests
+        run: python -m py.test
+      - name: setup asv
+        run: python write_asv_machine.py && asv check
+      - name: quick asv build of HEAD
+        run: asv run --quick
```
Comment (efiop) [inline .github/workflows/pull_request_build.yml:26]:
```suggestion
        run: asv run --quick

```

File: .github/workflows/main.yml (hunk 3 above)
Comment (pared) [reply .github/workflows/main.yml:85]: > Are they on the same fs?

Not sure, would need to debug.

They have few MB each, so I think there is no need to optimize them for now.

File: requirements.txt [hunk 6]
```diff
@@ -1,7 +1,7 @@
 virtualenv==16.7.10
 asv==0.4.1
 gitpython
-dvc==0.94.0
+dvc[s3]==0.94.0
```
Comment (efiop) [inline requirements.txt:4]: Aah, right, here is the answer to that run-cache question. We are still using 0.94.0 here. Ok, let's not switch to 1.0 for now. We'll do that after the official realeas.

File: .github/workflows/main.yml (hunk 4 above)
Comment (pared) [reply .github/workflows/main.yml:87]: I did not, will look into that soon.

File: .github/workflows/main.yml (hunk 2 above)
Comment (efiop) [reply .github/workflows/main.yml:10]: Oh, so we don't even need to create a bot user? Github Actions are really a box of treats, aren't they? :slightly_smiling_face:

File: .github/workflows/main.yml (hunk 3 above)
Comment (efiop) [reply .github/workflows/main.yml:85]: Got it. Let's leave as is then. Thanks!

File: .github/workflows/main.yml (hunk 2 above)
Comment (pared) [reply .github/workflows/main.yml:10]: yep, cron jobs are also in

**Comment by efiop** [comment]
> "quick build" for pull_requests - PR's from outside of dvc-bench repo, are not allowed to have the repo secrets -> no AWS credentials, so actual build should happen after push to dvc-bench, we might want to consider "pre-master" branch that would take care of building, so that we are sure master has only "good" changes

@pared So when people will submit a new benchmark PR, we won't run them until we merge? So we could merge broken benchmarks and won't even notice it until the master build? Or do I misunderstand it?

**Comment by pared** [comment]
@efiop So what we are doing after this change is, upon pull request, run `asv run --quick` which is equal to:
- running each benchmark only once (vs 3-5 that we do now)
- running it only for `dvc`'s HEAD

As of now I intend it as a `fast check` because with no access to remote, there is no point in running all benchmarks - we won't be able to download and (later) push them.

Alternative would be to have our results in public remote.

**Comment by efiop** [comment]
@pared 

> running each benchmark only once (vs 3-5 that we do now)

:+1: 

> running it only for dvc's HEAD

This might be problematic, because one of the benchmarks might try to use some features that are not available in older versions of dvc and we will only learn about it once we merge :slightly_frowning_face: 

> As of now I intend it as a fast check because with no access to remote, there is no point in running all benchmarks - we won't be able to download and (later) push them.

By remote you mean the remote that we can upload the results to? Well, but in that case it is better to have 1 excessive run while reviewing the PR than to fix the broken master. Also, we can provide read-only access to the remote (e.g. through http like we do in core dvc) so that PR builds could leverage it and not run everything. And we also have github artifacts, which (IIRC) could be used to upload the PR results too without any special access, so we could then potentially leverage its results to not build again on master.

**Comment by efiop** [comment]
@pared But again, the quickcheck will do the trick for now, no need to solve public read access in this particular PR, could totally do that on top :+1:

**Comment by pared** [comment]
@efiop merging current change and creating a few new ones:
https://github.com/iterative/dvc-bench/issues/61
https://github.com/iterative/dvc-bench/issues/62

File: .github/workflows/main.yml (hunk 2 above)
Comment (casperdcl) [reply .github/workflows/main.yml:10]: If you push using the builtin github token then it seems a job shouldn't be triggered - https://docs.github.com/en/actions/reference/events-that-trigger-workflows#triggering-new-workflows-using-a-personal-access-token

File: .github/workflows/main.yml (hunk 4 above)
Comment (casperdcl) [reply .github/workflows/main.yml:87]: is there a reason for using dvc rather than the builtin action cache mechanism?

Comment (efiop) [reply .github/workflows/main.yml:87]: @casperdcl Being able to pull(and even possibly push) results locally on our own machines.

**Pull request merged**